multi_cgra/test/MeshMultiCgraTemplateRTL_test.py

```diff
@@ -0,0 +1,626 @@
+    links[4].dstPort = PORT_WEST
+    links[4].validatePorts()
+
+    links[5].srcTile = tiles[0][1]
+    links[5].dstTile = tiles[0][0]
+    links[5].srcPort = PORT_WEST
+    links[5].dstPort = PORT_EAST
+    links[5].validatePorts()
+
+    links[6].srcTile = tiles[1][0]
+    links[6].dstTile = tiles[1][1]
+    links[6].srcPort = PORT_EAST
+    links[6].dstPort = PORT_WEST
+    links[6].validatePorts()
+
+    links[7].srcTile = tiles[1][1]
+    links[7].dstTile = tiles[1][0]
+    links[7].srcPort = PORT_WEST
+    links[7].dstPort = PORT_EAST
+    links[7].validatePorts()
+
+    links[8].srcTile = tiles[0][0]
+    links[8].dstTile = tiles[1][0]
+    links[8].srcPort = PORT_NORTH
+    links[8].dstPort = PORT_SOUTH
+    links[8].validatePorts()
+
+    links[9].srcTile = tiles[1][0]
+    links[9].dstTile = tiles[0][0]
+    links[9].srcPort = PORT_SOUTH
+    links[9].dstPort = PORT_NORTH
+    links[9].validatePorts()
+
+    links[10].srcTile = tiles[0][1]
+    links[10].dstTile = tiles[1][1]
+    links[10].srcPort = PORT_NORTH
+    links[10].dstPort = PORT_SOUTH
+    links[10].validatePorts()
+
+    links[11].srcTile = tiles[1][1]
+    links[11].dstTile = tiles[0][1]
+    links[11].srcPort = PORT_SOUTH
+    links[11].dstPort = PORT_NORTH
+    links[11].validatePorts()
+
+    links[12].srcTile = tiles[0][0]
+    links[12].dstTile = tiles[1][1]
+    links[12].srcPort = PORT_NORTHEAST
+    links[12].dstPort = PORT_SOUTHWEST
+    links[12].validatePorts()
+
+    links[13].srcTile = tiles[1][1]
+    links[13].dstTile = tiles[0][0]
+    links[13].srcPort = PORT_SOUTHWEST
+    links[13].dstPort = PORT_NORTHEAST
+    links[13].validatePorts()
+
+    links[14].srcTile = tiles[0][1]
+    links[14].dstTile = tiles[1][0]
+    links[14].srcPort = PORT_NORTHWEST
+    links[14].dstPort = PORT_SOUTHEAST
+    links[14].validatePorts()
+
+    links[15].srcTile = tiles[1][0]
+    links[15].dstTile = tiles[0][1]
+    links[15].srcPort = PORT_SOUTHEAST
+    links[15].dstPort = PORT_NORTHWEST
+    links[15].validatePorts()
+
+    def handleReshape( t_tiles ):
+      tiles = []
+      for row in t_tiles:
+        for t in row:
+          tiles.append(t)
+      return tiles
+
+    def safe_remove_port(tile, port):
+      tile_out_port_set = tile.invalidOutPorts
+      tile_in_port_set = tile.invalidInPorts
+      if port in tile_out_port_set:
+        tile_out_port_set.remove(port)
+      if port in tile_in_port_set:
+        tile_in_port_set.remove(port)
+
+    tiles_1 = copy.deepcopy(tiles)
+    tiles_2 = copy.deepcopy(tiles)
+    tiles_3 = copy.deepcopy(tiles)
+
+    # remove invalid ports for each tile on boundary
+    # cgra 0
+    print(f">>>>>>>>>>>>>>>> tiles[1][0] invalidOutPorts: {tiles[1][0].invalidOutPorts}, invalidInPorts: {tiles[1][0].invalidInPorts}")
+    safe_remove_port(tiles[1][0], PORT_NORTH)
+    safe_remove_port(tiles[1][1], PORT_NORTH)
+    safe_remove_port(tiles[0][1], PORT_EAST)
+    safe_remove_port(tiles[1][1], PORT_EAST)
```

Comment:
Why these ports are originally be categorized as `invalid`, because of the tile's construct?

Can we make these 4 `safe_remove_port()` be wrapped as a helper function called `keep_port_valid_on_boundary()`.